Update developer docs (#9787)
emscripten-core/emscripten: site/source/docs/contributing/developers_guide.rst

Developer's Guide
=================

> This article provides information that is relevant to people that contribute

Reviewer: ```suggestion
This article provides information that is relevant to people who want to contribute
```

> to Emscripten. We welcome contributions from anyone that is interested to

Reviewer: ```suggestion
to Emscripten. We welcome contributions from anyone that is interested in
```

> help out!

Reviewer: ```suggestion
helping out!
```

.. tip:: The information will be less relevant if you're just using Emscripten, but may still be of interest.

Getting Binaries
================

> For contributing to core Emscripten code, like ``emcc.py``, you don't need to

Reviewer: ```suggestion
For contributing to core Emscripten code such as ``emcc.py``, you don't need to
```

build any binaries as ``emcc.py`` is in Python, and the core JS generation is
in JavaScript. You do still need binaries for LLVM and Binaryen, which you can

> get using the emsdk,

Reviewer: ```suggestion
get using the emsdk:
```

::

    emsdk install tot-upstream
    emsdk activate tot-upstream

That gets a "tip-of-tree" build of the very latest binaries. You can use those
binaries with a checkout of the core Emscripten repository, simply by calling
``emcc.py`` from that checkout, and it will use the binaries from the emsdk.

If you do want to contribute to LLVM or Binaryen, or to test modifications
to them, you can 
:ref:`build them from source <installing-from-source>`.

Repositories and branches of interest
=====================================

The Emscripten main repository is https://github.com/emscripten-core/emscripten.

> The main branch is **incoming**, which exists for historic reasons as we did

Reviewer: Side note, I wonder if nowish (or perhaps the fastcomp fork) is a good time to switch to master as the tip of tree.

Author: I think now is as good as any time. I'd like to do it when I have the time (maybe after debug info?).

  an existing one. If you're not sure how to test your code, feel free to ask
  for help.
- We normally squash and merge PRs, which means the PR turns into a single
  commit on the target branch. Because of that, it's ok to have merge commits
  in the PR itself, as they get removed. Please put a good description for
  the final commit in the PR description, and we'll use it when squashing.

Code reviews
============

One of the core developers will review a pull request before merging it. If
several days pass without any comments on your PR, please comment in the PR
which will ping them. (If that happens, sorry! Sometimes things get missed.)

Compiler overview
=================

The :ref:`Emscripten Compiler Frontend (emcc) <emccdoc>` is a python script that manages the entire compilation process:

- **emcc** calls :term:`Clang` to compile C++ and ``wasm-ld`` to link it. It
  builds and integrates with the Emscripten system libraries, both the
  compiled ones and the ones implemented in JS.
- **emcc** then calls `emscripten.py <https://github.com/emscripten-core/emscripten/blob/master/emscripten.py>`_
  which performs the final transformation to wasm (including invoking

>   **wasm-emscripten-finalize** from Binaryen) and calls the JS compiler

Reviewer 2: Is wasm-emscripten-finalize described somewhere? If so, it would be useful to link to it, if not, then a short high-level description might be helpful too.

Author: It's an internal tool really, I don't think we have any more docs than this page. Maybe the tools `--help` output has a little more info...

I can add some description here, but not sure what kind of text you're thinking of?

Reviewer 2: I could try to make a suggestion, but tbh I don't know what it does myself, hence the suggestion to add some description for all potential contributors :) "Extracts metadata and generates JavaScript bindings"? (again, I have no idea if this is what it actually does)

  (see ``src/compiler.js`` and related files) which emits the JS.

> - If optimizing wasm, **emcc** will then call **wasm-opt**, run meta-dce, and

Reviewer 2: Same for meta-dce (I *think* it was documented somewhere?).

Author: there is a high-level description of the entire process here: https://hacks.mozilla.org/2018/01/shrinking-webassembly-and-javascript-code-sizes-in-emscripten/ is that what you mean?

Reviewer 2: Yeah, that looks good as a reference link. It just helps to assume no previous knowledge on parts of the process from the reader IMO.

- Make pull requests to **incoming**, not master.

> - Do not include merge commits in pull requests; include only commits with the new relevant code.

Reviewer: You could mention that our default github merge method is "squash and merge". Or maybe that makes this statement unnecessary? not sure.

Author: Oh right, this is out of date given we squash+merge - it's ok to have merge commits given that. I reworded and applied the other feedback.
